_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+access. When compiling code for space optimizations (using `-Os` CFLAG), I would have expected unaligned
+accesses to get optimized using this instruction. What I noticed is for some reason if the
+unaligned store is through a packed structure, the optimization does not take place.
+
+### pyelftools
+
+Pyelftools is a pure python package capable of parsing ELF and DWARF information. If you are
+looking to perform your own analysis on ELF files, I'd highly recommend checking out the
+repo. There is an [examples](https://github.com/eliben/pyelftools/tree/master/examples) folder in
+the project that you can use as inspiration to get started.
+
+#### Pure Python GNU Binutils size Command Equivalent
+
+As a quick example, we can implement an equivalent of the `size` binutil command in about 30 lines
+of python!
+
+```python
+# pyelftools_size.py
+def _get_allocated_section_sizes(elffile):
+    from elftools.elf.constants import SH_FLAGS
+
+    text = data = bss = 0
+    for section in elffile.iter_sections():
+        sh_flags = section["sh_flags"]
+        if sh_flags & SH_FLAGS.SHF_ALLOC == 0:
+            continue  # Section is not in final binary
+
+        section_size = section["sh_size"]
+
+        if (sh_flags & SH_FLAGS.SHF_EXECINSTR != 0 or
+            sh_flags & SH_FLAGS.SHF_WRITE == 0):
+            text += section_size
+        elif section["sh_type"] != "SHT_NOBITS":
+            data += section_size
+        else:
+            bss += section_size
+
+    return (text, data, bss)
+
+
+def size_elf(fname):
+    from elftools.elf.elffile import ELFFile
+
+    with open(fname, "rb") as ef:
+        text, data, bss = _get_allocated_section_sizes(ELFFile(ef))
+    print(f"   text\t   data\t    bss\t    dec\t    hex\tfilename")
+    tot = text + data + bss
+    print(f"{text:7d}\t{data:7d}\t{bss:7d}\t{tot:7d}\t{tot:7x}\t{fname")
+
+
+if __name__ == "__main__":
+    import sys
+
+    if len(sys.argv) < 2:
+        print("Please specify a path to an ELF")
+        exit(-1)
+    size_elf(sys.argv[1])
+
+```
+
+We can compare the binutil size to the python size to confirm it works!
+
+```bash
+$ python pyelftools_size.py nrf52_example/build/nrf52.elf
+   text    data     bss     dec     hex	filename
+   6340     244   11684   18268    475c	nrf52_example/build/nrf52.elf
+$ arm-none-eabi-size nrf52_example/build/nrf52.elf
+   text    data     bss     dec     hex	filename
+   6340     244   11684   18268    475c	nrf52_example/build/nrf52.elf
+```
+
+### LLVM
+
+The LLVM project has [clones](https://llvm.org/docs/CommandGuide/index.html#gnu-binutils-replacements) of most of the binutil commands we discussed above as well as a
+bunch of tools of its own. There's a healthy competition between the LLVM and GNU projects which
+means if a new feature makes it to one of them, there's typically work already underway to bring
+the other to parity. If you haven't looked at LLVM before, it's definitely worth checking out!
+
+### abidiff
```

Comment:
I'd move this one higher, it's probably something more people have not heard about.